File: app/utils/dawproject_python/markers.py
from lxml import etree as ET
from classes.timeline import Timeline
from classes.marker import Marker


class Markers(Timeline):
    XML_TAG = "Markers"

    def __init__(
        self,
        markers=None,
        track=None,
        time_unit=None,
        name=None,
        color=None,
        comment=None,
    ):
        super().__init__(track, time_unit, name, color, comment)
        self.markers = markers if markers else []

    def to_xml(self):
        markers_elem = ET.Element(self.XML_TAG)

        if self.time_unit:
            markers_elem.set("timeUnit", str(self.time_unit))
        if self.track:
            markers_elem.set(
                "track", str(self.track.id)
            )  # Assuming track has an id attribute

        for marker in self.markers:
            markers_elem.append(marker.to_xml())

        if self.name:
            markers_elem.set("name", self.name)
        if self.color:
            markers_elem.set("color", self.color)
        if self.comment:
            markers_elem.set("comment", self.comment)

        return markers_elem

    @classmethod
    def from_xml(cls, element):
        instance = super().from_xml(element)

        markers = []
        for marker_elem in element.findall("Marker"):
            markers.append(Marker.from_xml(marker_elem))
        instance.markers = markers

        return instance

File: app/utils/dawproject-new/markers.ts
// dawproject/markers.ts
import { Marker } from "./marker";
import type {
  Markers as MarkersType,
  TimeUnit,
  XsString,
} from "./project-schema";
import { Timeline } from "./timeline";

/**
 * Represents a collection of markers, typically within an Arrangement or other timeline.
 * Corresponds to the 'markers' complex type in Project.xsd.
 * Inherits attributes and child elements from Timeline.
 */
export class Markers extends Timeline implements MarkersType {
  // Property corresponding to child elements

  /**
   * A collection of marker points on the timeline.
   * (Required child element - unbounded)
   */
  public Marker: Marker[] = []; // Initialized as empty array for unbounded element

  /**
   * @param timeUnit - The time unit used for the timeline. (Optional attribute inherited from Timeline)
   * @param track - A reference to the track this timeline belongs to. (Optional attribute inherited from Timeline - xs:IDREF)
   * @param name - The name of the markers collection. (Optional attribute inherited from Nameable)
   * @param color - The color of the markers collection. (Optional attribute inherited from Nameable)
   * @param comment - A comment for the markers collection. (Optional attribute inherited from Nameable)
   */
  constructor(
    timeUnit: TimeUnit = "beats",
    track?: XsString,
    name?: XsString,
    color?: XsString,
    comment?: XsString
  ) {
    super(timeUnit, track, name, color, comment);
  }
}
